### PR TITLE
fix #1051: resource validation is slow

### LINKED ISSUE #1051: new validation seems quite slow
now we've validation builds seem quite slow. e.g. if you clone and build this repo it used to be pretty fast: https://github.com/fabric8io/fabric8-platform

Now its about 9 minutes on my laptop - just doing YAML generation/validation mostly. After 1 build there should be no downloading from the internet or anything; so should be pretty snappy. 

I wonder if the validation stuff could be tuned a bit? e.g. is it loading the schema from scratch as it validates every YAML file in every repo?

### PR DESCRIPTION
**Pull request by hrishin**
Fixed #1051 
- added JSON schema for validation https://github.com/fabric8io/kubernetes-model/issues/229
- replaced Github hosted schema with offline schema

### CODE REVIEW

**Comment by nicolaferraro** [comment]
Great @hrishin, I'll review it soon!

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java [hunk 1]
```diff
@@ -66,6 +66,7 @@
 import java.io.File;
 import java.io.FileFilter;
 import java.io.IOException;
+import java.net.URISyntaxException;
```
Comment (nicolaferraro) [inline plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Not needed

File: samples/spring-boot/pom.xml [hunk 2]
```diff
@@ -27,7 +27,7 @@
     <parent>
         <groupId>org.springframework.boot</groupId>
         <artifactId>spring-boot-starter-parent</artifactId>
-        <version>1.4.1.RELEASE</version>
+        <version>RELEASE</version>
```
Comment (nicolaferraro) [inline samples/spring-boot/pom.xml]: Typo

File: core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java [hunk 3]
```diff
@@ -101,15 +92,21 @@ private void setupIgnoreRules(ResourceClassifier target) {
      * @throws ConstraintViolationException
      * @throws IOException
      */
-    public int validate() throws ConstraintViolationException, IOException {
-        for (File resource : resources) {
+    public int validate() throws IOException, URISyntaxException {
```
Comment (nicolaferraro) [inline core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java]: Need to fix also `@throws` declarations in comments

File: core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java [hunk 4]
```diff
@@ -101,15 +92,21 @@ private void setupIgnoreRules(ResourceClassifier target) {
...
+        for(File resource: resources) {
             if (resource.isFile() && resource.exists()) {
-                log.info("validating %s resource", resource.toString());
-                JsonNode resourceNode = geFileContent(resource);
-                JsonSchema schema = getJsonSchema(prepareSchemaUrl(resourceNode));
-
-                Set<ValidationMessage> errors = schema.validate(resourceNode);
-                processErrors(errors, resource);
+                try {
+                    log.info("validating %s resource", resource.toString());
```
[the remark this answers is not in the document]
Comment (hrishin) [reply core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java]: Yes @nicolaferraro we could but we kept it info messages intentionally which all resource plugin is validating.

File: core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java (hunk 3 above)
Comment (hrishin) [reply core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java]: fixed :+1:

File: core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java [hunk 5]
```diff
@@ -151,13 +148,29 @@ private String getErrorMessage(File resource, Set<ConstraintViolationImpl> viola
...
-        return String.format(jsonSchemaPath, target, "master", resourceNode.get("kind").toString().toLowerCase()).replace("\"", "");
+    private void getResourceProperties(String kind, JSONObject jsonSchema) {
+        jsonSchema.put("properties" , jsonSchema.getJSONObject("resources")
+                .getJSONObject(kind.replaceAll("\"", "").toLowerCase())
+                .getJSONObject("properties"));
+    }
+
+    private void checkIfKindPropertyExists(String kind) {
+        if(kind == null) {
+            throw new JSONException("Invalid kind of resource or \'kind\' is missing from resource definition");
```
[the remark this answers is not in the document]
Comment (hrishin) [reply core/src/main/java/io/fabric8/maven/core/util/validator/ResourceValidator.java]: fixed :+1:

File: plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java (hunk 1 above)
Comment (hrishin) [reply plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java]: Are you sure?
Its used in the code

File: samples/spring-boot/pom.xml (hunk 2 above)
Comment (hrishin) [reply samples/spring-boot/pom.xml]: fixed :+1:

**Comment by nicolaferraro** [comment]
@hrishin looks good. Can we remove the json file before merging or it is used. I think the one used for validation is included in kubernetes-model, right?

**Comment by hrishin** [comment]
@nicolaferraro Thanks for review :)
Yes, we've removed it.

**Pull request merged**